lms/djangoapps/ccx/utils.py

```diff
@@ -0,0 +1,294 @@
+            'ccx/unenroll_email_allowedmessage.txt'
+        ),
+        'enrolled_unenroll': (
+            'ccx/unenroll_email_subject.txt',
+            'ccx/unenroll_email_enrolledmessage.txt'
+        ),
+    }
+
+    subject_template, message_template = email_template_dict.get(
+        message_type, (None, None)
+    )
+    if subject_template is not None and message_template is not None:
+        subject = render_to_string(subject_template, param_dict)
+        message = render_to_string(message_template, param_dict)
+
+    if subject and message:
+        message = message.strip()
+
+        subject = ''.join(subject.splitlines())
+        from_address = microsite.get_value(
+            'email_from_address',
+            settings.DEFAULT_FROM_EMAIL
+        )
+
+        send_mail(
+            subject,
+            message,
+            from_address,
+            [student],
+            fail_silently=False
+        )
+
+
+def get_all_ccx_for_user(user):
+    """return all CCXS to which the user is registered
+
+    Returns a list of dicts: {
+        ccx_name: <formatted title of CCX course>
+        ccx_url: <url to view this CCX>
+        ccx_active: True if this ccx is currently the 'active' one
+        mooc_name: <formatted title of the MOOC course for this CCX>
+        mooc_url: <url to view this MOOC>
+    }
+    """
+    if user.is_anonymous():
+        return []
+    current_active_ccx = get_current_ccx()
+    memberships = []
+    for membership in CcxMembership.memberships_for_user(user):
+        course = get_course_by_id(membership.ccx.course_id)
+        ccx = membership.ccx
+        ccx_title = ccx.display_name
+        mooc_title = get_course_about_section(course, 'title')
+        url = reverse(
+            'switch_active_ccx',
+            args=[course.id.to_deprecated_string(), membership.ccx.id]
+        )
+        mooc_url = reverse(
+            'switch_active_ccx',
+            args=[course.id.to_deprecated_string(), ]
+        )
+        memberships.append({
+            'ccx_name': ccx_title,
+            'ccx_url': url,
+            'active': membership.ccx == current_active_ccx,
+            'mooc_name': mooc_title,
+            'mooc_url': mooc_url,
+        })
+    return memberships
+
+def get_ccx_membership_triplets(user, course_org_filter, org_filter_out_set):
+    """
+    Get the relevant set of (CustomCourseForEdX, CcxMembership, Course)
+    triplets to be displayed on a student's dashboard.
+    """
+    # only active memberships for now
+    for membership in CcxMembership.memberships_for_user(user):
+        ccx = membership.ccx
+        store = modulestore()
+        with store.bulk_operations(ccx.course_id):
+            course = store.get_course(ccx.course_id)
+            if course and not isinstance(course, ErrorDescriptor):
+                # if we are in a Microsite, then filter out anything that is not
+                # attributed (by ORG) to that Microsite
```

Comment:
This is the unfortunate kind of NxM logic that we need to try to avoid.  I don't see a way to avoid it here, but this is why features are scrutinized so carefully to see if we can not spread the logic all over the codebase.